node/network/collator-protocol/src/validator_side.rs

```diff
@@ -0,0 +1,1369 @@
+			}
+		}
+
+		return Ok(());
+	}
+
+	let request_id = state.next_request_id;
+	state.next_request_id += 1;
+
+	let (tx, rx) = oneshot::channel();
+
+	let per_request = PerRequest {
+		received: tx,
+		result,
+	};
+
+	let request = CollationRequest {
+		received: rx,
+		timeout: Delay::new(state.request_timeout),
+		request_id,
+	};
+
+	state.requested_collations.insert((relay_parent, para_id.clone(), peer_id.clone()), request_id);
+
+	state.requests_info.insert(request_id, per_request);
+
+	state.requests_in_progress.push(request);
+
+	let wire_message = protocol_v1::CollatorProtocolMessage::RequestCollation(
+		request_id,
+		relay_parent,
+		para_id,
+	);
+
+	ctx.send_message(AllMessages::NetworkBridge(
+		NetworkBridgeMessage::SendCollationMessage(
+			vec![peer_id],
+			protocol_v1::CollationProtocol::CollatorProtocol(wire_message),
+		)
+	)).await?;
+
+	Ok(())
+}
+
+/// Networking message has been received.
+async fn process_incoming_peer_message<Context>(
+	ctx: &mut Context,
+	state: &mut State,
+	origin: PeerId,
+	msg: protocol_v1::CollatorProtocolMessage,
+)-> Result<()>
+where
+	Context: SubsystemContext<Message = CollatorProtocolMessage>
+{
+	use protocol_v1::CollatorProtocolMessage::*;
+
+	match msg {
+	    Declare(id) => {
+			state.known_collators.insert(origin.clone(), id);
+			state.peer_views.entry(origin).or_default();
+		}
+	    AdvertiseCollation(relay_parent, para_id) => {
+			state.advertisments.entry(origin.clone()).or_default().insert((para_id, relay_parent));
+			request_collation(ctx, state, relay_parent, para_id, origin, None).await?;
+		}
+	    RequestCollation(_, _, _) => {
+			// This is a validator side of the protocol, collation requests are not expected here.
+			return modify_reputation(ctx, origin, COST_UNEXPECTED_MESSAGE).await;
+		}
+	    Collation(request_id, receipt, pov) => {
+			received_collation(ctx, state, origin, request_id, receipt, pov).await?;
+		}
+	}
+
+	Ok(())
+}
+
+/// A leaf has become inactive so we want to
+///   - Cancel all ongoing collation requests that are on top of that leaf.
+///   - Remove all stored collations relevant to that leaf.
+async fn remove_relay_parent(
+	state: &mut State,
+	relay_parent: Hash,
+) -> Result<()> {
+	let mut remove_these = Vec::new();
+
+	state.requested_collations.retain(|k, v| {
+		if k.0 == relay_parent {
+			remove_these.push(*v);
+			false
+		} else {
+			true
+		}
+	});
+
+	for id in remove_these.into_iter() {
+		if let Some(info) = state.requests_info.remove(&id) {
+			info.received.send(()).map_err(|_| oneshot::Canceled)?;
+		}
+	}
+
+	state.collations.retain(|k, _| {
+		if k.0 == relay_parent {
+			false
+		} else {
+			true
+		}
+	});
+
+	Ok(())
+}
+
+/// Our view has changed.
+async fn handle_our_view_change(
+	state: &mut State,
+	view: View,
+) -> Result<()> {
+	let old_view = std::mem::replace(&mut (state.view), view);
+
+	let cloned_view = state.view.clone();
+	let removed = old_view.difference(&cloned_view).collect::<Vec<_>>();
```

Comment:
Why do we need `cloned_view`, if we just take a reference?

```suggestion
	let removed = old_view.difference(&state.view).collect::<Vec<_>>();
```